### ISSUE TITLE
3.1 Auto fixtures update or create schema.lock file

### ISSUE THREAD

**Issue by Laykou** [cakephp/cakephp#5937]
It would be helpful to automatically update fixtures `$fields` after the migration. The fixtures cannot be $imported when running on the third-party testing tool like codeship, travis, appveyor...

In Rails there is a `schema.rb` file which holds the current db schema and is loaded before testing. The best solution would be to have some `schema.lock` file which could be loaded running the command `cake tests prepare` or something like that, so that user can schedule when to dump the schema.

The `schema.lock` file would be automatically recreated whenever running `cake migrations migrate`.

I would use this feature with the [cake/codeception](http://github.com/cakephp/codeception) acceptance tests so it could be somehow connected so that the schema is automatically loaded before the acceptance tests.

Maintaining the fixtures and migrations isn't the best way for clean code :)
## TL;DR

Jump to the comment https://github.com/cakephp/cakephp/issues/5937#issuecomment-77759151. I created the [Schema plugin for CakePHP 3.0](https://github.com/Laykou/cakephp-schema)

**Comment by Laykou** [comment]
Another problem with fixtures is that they are executed one by one. Having the table `articles` with the `last_comment_id` column and `comments` with the `article_id` column is problem to create because the tables references each other (if the foreign key constraint is enabled). So the tables must be created at once.

This could be solved with the `schema.lock` file.

**Comment by jadb** [comment]
I had a few unfruitful attempts to try having fixtures and migrations play better together but never tried the lock idea you are suggesting @Laykou, which I kinda like.

Let's see what others think.

**Comment by Laykou** [comment]
In rails the command is `rake db:schema:load` and it can be used any time to load the entire db schema from the file. It should be allowed on the production table too, but very carefully. Will it delete the data or just somehow update? What if the schema is about to delete some columns? If the db is empty, everythin is ok. It could hold the last migration "tag" when the `schema.lock` file was generated and show some alert when trying to dump older schema than current db.

In Cake the command could be
- `cake db_schema_load`
- `cake db_load`

I think it should be supported by the core. Or should this be part of the [migration](http://github.com/cakephp/migrations) plugin?

**Comment by markstory** [comment]
One problem with a static schema file is that it most SQL table definitions are not portable between RDBMS. While this generally isn't a problem for applications, schema portability is important for CakePHP and plugins.

Isn't a static schema and data dump file just another way of expressing a database fixture?

**Comment by Laykou** [comment]
Yes it is. Therefore I think it will not be problem. It could have the same behavior backend as the Fixtures: columns + indexes + constraints.

However there would be these benefits:
- no `dump.sql` file which could be RDBMS specific
- load all tables in one transaction (solves problem with tables referencing each other)
- holds the migration stamp - can popup warning if trying to load older schema
- can be automatically recreated after each migration - updating fixtures `$fields` is not a good solution because developers can edit fixtures but the `schema.lock` file would be system only
- can have command line command to load the dump
- possibly can compare the current database and the `schema.lock` file
- can be used in testing but also to quickly load up the database to the last state (no need to go through all migrations). Sometimes you start the project with some database and then move it to the CakePHP and you don't have the whole schema built in migrations. So you need to copy the schema between RDBMS if you want to move to another computer.


Because usually your project is large with many dependencies and you usually want to load more fixtures if not all, because of constraints and relationships. This also brings up another problem: *loading all data from fixtures in one transaction*.
But I see that other frameworks like Ruby has this schema file, so it might not be such a wrong idea. :)

So the `schema.lock` file would be just one big array: as if you put all fixtures `$fields` together + the migration timestamp.

**Comment by markstory** [comment]
Rails' fixtures don't include data in the schema dump. Instead data is in yml files that are loaded in a similar way to ours. Separate fixture files are nice as tests can more easily ensure that only the data they need is loaded and the test runner does not spend time loading data that is not relevant to the test being run.

While I can see the advantages of having a schema dump I feel having two ways to define schema is not ideal. Have you considered if running the migrations against the test database before running tests would be another way to avoid duplicating schema data? We do this at my day job for a non-cake app and it solves duplicate schema definition.

**Comment by Laykou** [comment]
Ok, maybe the migrations are enough.. But you will not need to maintaine two schema definitions because the `schema.lock` file would be **system only** and you should not edit it. But I see what you are afraid of..... 
Currently the fixtures allows you to define different schema from migrations and you need to maintaine the migrations and fixtures.. Ok I know about *$import* and *running migrations* as you described. The `schema.lock` was just an idea of extracting the $fields from fixtures into separated file.

---
My problem is that I started the project **before** moving to the Cake. So I copied the database, built my models, but I **do not have initial migrations** for creating the tables. This might be also problem for some people, who decide to move to the amazing CakePHP from other frameworks... And the idea of schema file looked really clean to me.

---
I'm leaving it up to you. I just thought that schema and data should be **separated** from the fixtures as the schema is something which should be changed automatically. And usually in large projects there are many relationships among tables. In my case I need to *load almost all fixtures* because of the *foreign key* constraits so the benefit of loading only specific fixtures is lost.

And I think that creating empty schema doesn't take much time. I understand that you want to load only needed data, that's fine..... Maybe the fixtures could have an option to load the specific schema from the `schema.lock` file instead of the `$fields`. That way the fixtures would work as usual (you are able to load only specific tables) but you have also the benefit of the `schema.lock` file.

**Comment by markstory** [comment]
I totally get wanting to manage schema for a big project and I think having a schema file is a good option, we just need to figure out ways to handle edge cases and portability issues. Perhaps we don't have to handle the portability issue if we position a `schema.sql` file as an optimization when you are only working with a single rdbms.  I don't know how we would indicate to fixtures that their schema should already exist though.

**Comment by Laykou** [comment]
Hi,

so I have created this [Schema plugin for CakePHP 3.0](https://github.com/Laykou/cakephp-schema) which does what I was talking about. You can review it and if you want I'm willing to transfer the ownership to you to make it official CakePHP plugin or even integrate it into migrations plugin.

The only thing that is need to be done (except tests) is to trigger the schema save after the `cake migrations migrate`. Is there any possibility for a hook or something?

**Comment by markstory** [comment]
Exporting the abstract schema definitions is a neat idea. It avoids a bunch of the portability issues that SQL dump files like Rails uses have.

Loading this kind of schema file could be something that a FixtureManager does, which could close the loop on how to integrate a schema file with test suites.

Are there any events fired by Migrations that you could use to rebuild the schema file?

**Comment by Laykou** [comment]
Just for the record Rails uses abstract schema definition too. :)

But it cannot be loaded by the FixtureManager because fixture manager loads the whole table at once and I need to **firstly** create the table and after all tables are created the **foreign key constraints** can be applied. Therefore I need to do some manipulation with the `Cake\Database\Schema\Table` object, but it works.

And about the Migrations events: that was my question to you (or any member of the team) if there are any events fired?

**Comment by jadb** [comment]
There are events fired. Here's the one you were asking about, [afterMigrate](https://github.com/cakephp/migrations/blob/dd47d274b22799bd903ed069699e2c28a47e0d63/src/Command/Migrate.php#L58).

Nice plugin, I'll have to try it. Can you tell me a little more about seeding? Is it only from fixtures? Using the migrations files (with maybe environment defined flags)? Or can we support `config/Seed` like `config/Migration`?

**Comment by Laykou** [comment]
Currently I do not do any seeding. The only think plugin does is it stores current schema to the file and later loads it back. Only the schema with indexes and constraints (using the same array syntax as fixtures).

Currently it is a standalone plugin - doesn't do anything with fixtures or migrations.

**Comment by lorenzo** [comment]
very interesting @Laykou ! I would definitely be interested in merging that into the migrations plugin

**Comment by jadb** [comment]
Ok, in light of that @Laykou, and given your obvious experience in the matter, how would you suggest seeding takes place? And how much work you think it would require?

**Comment by lorenzo** [comment]
@Laykou A couple things about your plugin.

1. It should use templates instead of building the php file directly in the task. You could utilize the bake tasks for helping you render that.
2. I think the file should be named `schema.php` instead of `schema.lock` That would help prevent leakage of information in case the file is served by mistake in a browser.

**Comment by Laykou** [comment]
@lorenzo 
1. I just wasn't sure how can I use the bake. Or do you mean just to copy necessary code out of the Bake plugin to make the Schema plugin work with templates?
2. Yeah I was thinking about that. Agree

@jadb
I'm still not sure about seeding but it could be in the `config/Seed` folder - class-less files which would contain the seeding logic, or maybe some helpers could be available there for faster manipulating with object.

Another solution is to create (one) `yaml` file which would contain all tables. Because having separated fixtures is a good idea but usually tables have so many dependencies that you still need to load all the fixtures. And maintaining 30-50 fixture files isn't fun.

The Rails have one `seed.rb` file which contains all the logic. So maybe something like `config/seed.php` file would be enough.

@lorenzo What about renaming or better **aliasing** the migration plugin with `db`? You could run `cake db migrate`, `cake db save`, `cake db load`, `cake db seed` and `cake db create` to create the database if it doesn't exist.

**Comment by lorenzo** [comment]
@Laykou Check the migrations plugin, it makes use of the bake plugin for generating the first migration snapshot. Also, I don't think renaming the plugin is a bad idea, but perhaps it will introduce some pain for people that are already using the current plugin.

**Comment by lorenzo** [comment]
@Laykou Also, I'm derp and did not read that you suggested aliasing. That seem like a better option :D

**Comment by Laykou** [comment]
@lorenzo I've remake the schema saving to work via SimpleBakeTask as you suggested.

**Comment by Laykou** [comment]
@jadb I've just added the [seed functionality](https://github.com/Laykou/cakephp-schema#seed). It works with `config/seed.php` file which should return array of tables and their rows (see the readme) + you can work with CakeORM or custom libraries and functions. And the seed can also return empty array but the seeding can be done manually. So it's up to developer.

Seeding can be triggered by `cake schema seed`. I've tried to make IDENTITY_INSERT work on Sqlserver but it doesn't work (also cakephp's fixtures doesn't insert ids on SQL server).

@markstory The `config/seed.php` and the `config/schema.php` have the same format as fixtures. They are just condensed fixtures into one file. So maybe the fixtures could be setup to read from these files (similar as import from db works) and we can benefit from the fixtures functionality loading separated tables + have everyting on one place.

**Comment by Laykou** [comment]
I have managed to setup the `schema` + `seed` + `cakephp/codeception` + `acceptance tests` + `phantomjs` + `circleci and codeship`. Everything works automatically as expetected and is really easy to maintain (migrations automatically update schema and seeds are in one file). I can share my knowledge later - now I'm really tired but happy that something works :)

From now the refactoring is probably needed, tests and maybe merging into migration as @lorenzo suggested. @markstory @jadb thanks for great features and effort you've put into CakePHP.

**Comment by markstory** [comment]
Having a fixture mode for no schema sounds totally reasonable. When the first fixturized test is run if the schema.php file exists it could be loaded. Fixtures could then have `$schema = false` to tell the fixture that it's schema is managed elsewhere.

**Comment by Laykou** [comment]
Totally agree. Other enhancement could be **auto fixtures** when the `ArticlesFixture.php` doesn't exist but since we have `schema.php` and `seed.php` which contains the keys `articles` describing the schema and data for the table, the Fixture can be automatically built without any `ArticlesFixture.php` file. So the `public $fixtures = ['app.articles']` in TestCases would also work.

Since there is no `ArticlesFixtures.php` but there is `schema.php` the table can be loaded to the default connection. And if there are some data in `seed.php` the data can be loaded. And if the user needs to customize connection or add additional mehods, he can create the `ArticlesFixtures.php` and set `$schema = false` or `$records = false` to still benefit from those files.

This could remove duplicated code and handling many Fixture files but still keep them and make them work as they are.

**Comment by dereuromark** [comment]
Should this ticket be moved to migrations plugin? Sounds like the remaining issues not implemented there should also be solved in that plugin rather.

**Comment by dereuromark** [comment]
Can someone move this to migrations plugin?
